### ISSUE TITLE
A way to hide history (save and restore) commands from panel

### ISSUE THREAD

**Issue by informafix** [tmux-plugins/tmux-resurrect#278]
Hello

This is not an issue but a behavior i'd like to change.

When saving history, tmux-resurrect shows its command in each tmux panes

I've found in **SAVE.SH** line **158** `local history_w='fc -lLn -64 >'`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; also line **159** `local history_r='fc -R'`
and in **RESTORE.SH** line **285**  `local read_command="fc -R '$history_file'; clear"`

So I ask myself is there 's a way to hide those commands showing very frequently in tmux panes, like adding `> /dev/null` at the end of those commands ?

thanks

**Comment by leseb** [comment]
I have the exact same problem, all my panes are filled with `fc -lLn -64 > '/Users/leseb/.tmux/resurrect/zsh_history-leseb:7.0`. I didn't have this until recently and it's quite annoying :(

**Comment by vifreefly** [comment]
Yes, I found that annoying too. Even if it's not possible to do not print these command to the panes, it is possible to clear the pane output after command execution. For example I have following binding to clear window pane:
```
bind C-l send-keys -R \; send-keys C-l \; clear-history
```

**Comment by tonywoode** [comment]
everything is amazing with my resurrect life! Apart from this. I updated my tmux plugins after 'some time' today to find that these things happen at the continuum interval:

1. multiple `fc -R` commands appear in each terminal
2. some of those commands echo back things like:
`^E fc -R '/Users/twoode/.tmux/resurrect/zsh_history-0:1.2'`, so quickly take up the whole pane
3. the running of those commands messes up my history, so if i'm relying on repeated up-cursor running of the same command, I can't run it any longer, history has been altered in some way I can't quite grok
4. A large message regarding the `fc` command sometimes totally replaces my vim session, requiring multiple presses of the return key to clear and get back to all the vim windows
5. the tmux  command line sometimes gets activated with the message displayed `(jump forward)`, again requiring interaction to clear

I see this is a very similar issue to https://github.com/tmux-plugins/tmux-resurrect/issues/86, but with zsh. I also see this: https://github.com/AdrieanKhisbe/zsh-quiet-accept-line - Adriene says that he made this plugin to help ease this issue, but he doesn't say how he modded, I don't see his dot files nor interaction with the issues here...

Having these issues with interval at 15 minutes is not doable. I didn't even notice zsh wasn't saving my history previously. How do I deactivate the zsh-history? These lines in `scripts/save.sh` don't seem to guarded by `if`s, is there no current way to turn this off without commenting them?

```
	elif [ "$pane_command" = "zsh" ]; then
		# fc -W does not work with -L
		# fc -l format is different from what's written by fc -W
		# fc -R either reads the format produced by fc -W or considers
		# the entire line to be a command. That's why we need -n.
		# fc -l only list the last 16 items by default, I think 64 is more reasonable.
		local history_w='fc -lLn -64 >'
		local history_r='fc -R'
```

**Comment by bruno-** [comment]
Hi guys,

I'm sorry that saving shell history isn't working flawlessly for you. As you know, [this feature is marked "experimental"](https://github.com/tmux-plugins/tmux-resurrect/blob/master/docs/restoring_shell_history.md) so yea, not perfect.

I thought hard if this downside could be avoided (for bash) and found no solution. I personally gave up on using this feature and was considering removing it altogether because it cannot be made right. At this point, I'm at most accepting improvements for it.

I'd also be ok with honestly writing downsides upfront in this doc so that new users don't get too excited about it https://github.com/tmux-plugins/tmux-resurrect/blob/master/docs/restoring_shell_history.md

**Comment by nickjj** [comment]
@bruno-, if this feature is on its way to being abandoned, do you have any ideas on how to approach this problem with a different strategy all together?

For example, there's this fork of your project (https://github.com/markeissler/tmux-resurrect-ng#why) that says it integrates with your bash prompt directly and runs some commands in the background. I'd rather stick to your plugin because it's well maintained and works with newer versions of tmux.

**Comment by bruno-** [comment]
Hey,

I'm open to accepting a PR with another working strategy that's not too hacky.

**Comment by nickjj** [comment]
I'm new to the tmux ecosystem beyond just using it, so I wouldn't really know where to begin with that when it comes to plugin development. Do you have any general ideas on how it could be implemented in a non-hacky way?

Currently I'm running Windows 10 with WSL and tmux works great there, but Windows is Windows, which means quite a few of unplanned reboots (forced patches that are automated by Microsoft), so losing my sessions is a real problem. Your plugin is a fantastic fix for that, but history is important too.

**Comment by roachsinai** [comment]
Hi @bruno- . It seems once I set `set -g @resurrect-save-shell-history 'on'` on .tmux.conf, I have to set `set -g @resurrect-save-shell-history 'off'` to unset the saving history function which means comment  `set -g @resurrect-save-shell-history 'on'` on .tmux.conf doesn't work.

And after restore, I always got a `%` on the shell, though it's not an so annoying problem, I want to know it's my setting error or I got same with everyone.

![image](https://user-images.githubusercontent.com/9500049/50764759-b910b480-12ae-11e9-9ef7-98c4e6cc4a95.png)

**Comment by macandchief** [comment]
I'm not a long time tmux-resurrection user but really like it. History is kind of crucial for me, without history I suppose using a tiling wm with pre-configured terminals opening at start might an alternative. At least history would be available from the shell. Actually I intended to ask whether there is a chance to get history working with ksh but I suppose I know the answer. Anyhow, this is a very useful tool, thank you for your work.

**Comment by bruno-** [comment]
@macandchief history feature is very buggy, and honestly not recommended as it just does not work. I'd like to remove it from this plugin.

**Comment by liamaharon** [comment]
I've found another way to keep history of commands I make inside my tmux panes: 

https://askubuntu.com/a/339925
>You have to preserve bash history in multiple bash shells. To do this, be sure that you have the following lines in your ~/.bashrc file:
>```
># avoid duplicates..
>export HISTCONTROL=ignoredups:erasedups
>
># append history entries..
>shopt -s histappend
>
># After each command, save and reload history
>export PROMPT_COMMAND="history -a; history -c; history -r; $PROMPT_COMMAND"
>```

it's not perfect as all panes share the same history, but it's **much** better than nothing

hope this helps someone

**Comment by pdecat** [comment]
Maybe sending signals to trigger a bash helper function could make this work without invoking `history` commands from the terminal?

http://blog.nqzero.com/2016/02/save-bash-history-remotely-by-sending.html
